lib/os_io.c

```diff
@@ -37,6 +37,199 @@ ___io_module ___io_mod =
 
 /*---------------------------------------------------------------------------*/
 
+/* poll dynamic fdset memory management. */
+
+#ifdef USE_select_or_poll
+
+___HIDDEN int ___fdset_heap_overflow;
+
+___HIDDEN int ___fdset_init
+   ___P((___processor_state ps),
+        (ps)
+        ___processor_state ps;)
+{
+  void *readfds = NULL;
+  void *writefds = NULL;
+  void *exceptfds = NULL;
+
+  int size = MAX_CONDVARS;
+  int bytes;
+
+  if (size < 8 * sizeof (fd_set))
+    size = 8 * sizeof (fd_set);
+  bytes = ___CEILING_DIV (size, 8);
+  
+  readfds  = ___ALLOC_MEM (bytes);
+  if (!readfds)
+    goto error;
+
+  writefds = ___ALLOC_MEM (bytes);
+  if (!writefds)
+    goto error;
+
+#ifdef USE_select
+  exceptfds = ___ALLOC_MEM (bytes);
+  if (!exceptfds)
+    goto error;
+#endif
+
+  memset (readfds, 0, bytes);
+  memset (writefds, 0, bytes);
+#ifdef USE_select
+  memset (exceptfds, 0, bytes);
+#endif
+
+  ps->os.fdset.readfds = readfds;
+  ps->os.fdset.writefds = writefds;
+  ps->os.fdset.exceptfds = exceptfds;
+  ps->os.fdset.size = size;
+
+  return 0;
+
+ error:
+  if (readfds)
+    ___FREE_MEM (readfds);
+  if (writefds)
+    ___FREE_MEM (writefds);
+  if (exceptfds)
+    ___FREE_MEM (exceptfds);
+
+  return 1;
+}
+
+___HIDDEN int ___fdset_realloc
+   ___P((___processor_state ps,
+         int fd),
+        (ps,
+         fd)
+        ____processor_state ps;
+        int fd;)
+{
+  void *readfds = NULL;
+  void *writefds = NULL;
+  void *exceptfds = NULL;
+  int oldbytes, newbytes;
+  int oldsize = ps->os.fdset.size;
+  int newsize = oldsize;
+
+  while (newsize <= fd)
+    newsize = newsize * 2;
+
+  if (oldsize == newsize) /* size unchanged, no need to realloc */
+      return 0;
```

Comment:
This will need to be aligned as well.
I think we need a change to the interface to support the global vm size -- it should take a new target
size, and not try to compute the new size itself.
The sizing selection logic goes to the global state maintenance then.